dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -106,15 +105,22 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
         i3: TypedEncoder[Out]
       ): TypedDataset[Out] = {
 
-        val cols = columns.toList[UntypedExpression[T]].map(c => new Column(c.expr))
-
-        val selected = dataset.toDF()
-          .agg(cols.head.alias("_1"), cols.tail: _*)
-          .as[Out](TypedExpressionEncoder[Out])
-          .filter("_1 is not null") // otherwise spark produces List(null) for empty datasets
-
-        TypedDataset.create[Out](selected)
-      }
+      val underlyingColumns = columns.toList[UntypedExpression[T]]
+      val cols: Seq[Column] = for {
+        (c, i) <- columns.toList[UntypedExpression[T]].zipWithIndex
+      } yield new Column(c.expr).as(s"_${i+1}")
+
+      // Workaround to SPARK-20346. The alternative is to allow the result to be Vector(null) for empty DataFrames.
```

Comment:
`One alternative is to allow the result to be Vector(null) for empty DataFrames. Another one would be to return an Option`